File: kyototycoon/example/ktrestex.pl
use strict;
use warnings;
use LWP::UserAgent;
use URI::Escape;

{
    # RESTful interface of Kyoto Tycoon
    package KyotoTycoon;

    # constructor
    sub new {
        my $self = {};
        bless $self;
        return $self;
    }

    # connect to the server
    sub open {
        my ($self, $host, $port, $timeout) = @_;
        $host = "127.0.0.1" if (!defined($host));
        $port = 1978 if (!defined($port));
        $timeout = 30 if (!defined($timeout));
        $self->{base} = "http://$host:$port/";
        $self->{ua} = LWP::UserAgent->new(keep_alive => 1);
        $self->{ua}->timeout($timeout);
        return undef;
    }

    # close the connection
    sub close {
        my ($self) = @_;
        $self->{ua} = undef;
        return undef;
    }

    # store a record
    sub set {
        my ($self, $key, $value, $xt) = @_;
        my $url = $self->{base} . URI::Escape::uri_escape($key);
        my @headers;
        if (defined($xt)) {
            $xt = time() + $xt;
            push(@headers, "X-Kt-Xt");
            push(@headers, $xt);
        }
        my $req = HTTP::Request->new(PUT => $url, \@headers, $value);
        my $res = $self->{ua}->request($req);
        my $code = $res->code();
        return $code == 201;
    }

    # remove a record
    sub remove {
        my ($self, $key) = @_;
        my $url = $self->{base} . URI::Escape::uri_escape($key);
        my $req = HTTP::Request->new(DELETE => $url);
        my $res = $self->{ua}->request($req);
        my $code = $res->code();
        return $code == 204;
    }

    # retrieve the value of a record
    sub get {
        my ($self, $key) = @_;
        my $url = $self->{base} . URI::Escape::uri_escape($key);
        my $req = HTTP::Request->new(GET => $url);
        my $res = $self->{ua}->request($req);
        my $code = $res->code();
        return undef if ($code != 200);
        return $res->content();
    }

}

# sample usage
my $kt = new KyotoTycoon;
$kt->open("localhost", 1978);
$kt->set("japan", "tokyo", 30);
printf("%s\n", $kt->get("japan"));
$kt->remove("japan");
$kt->close();

File: kyototycoon/example/ktrestex.rb
require 'uri'
require 'net/http'

# RESTful interface of Kyoto Tycoon
class KyotoTycoon
  # connect to the server
  def open(host = "127.0.0.1", port = 1978, timeout = 30)
    @ua = Net::HTTP::new(host, port)
    @ua.read_timeout = timeout
    @ua.start
  end
  # close the connection
  def close
    @ua.finish
  end
  # store a record
  def set(key, value, xt = nil)
    key = "/" + URI::encode(key)
    req = Net::HTTP::Put::new(key)
    if xt
      xt = Time::now.to_i + xt
      req.add_field("X-Kt-Xt", xt)
    end
    res = @ua.request(req, value)
    res.code.to_i == 201
  end
  # remove a record
  def remove(key)
    key = "/" + URI::encode(key)
    req = Net::HTTP::Delete::new(key)
    res = @ua.request(req)
    res.code.to_i == 204
  end
  # retrieve the value of a record
  def get(key)
    key = "/" + URI::encode(key)
    req = Net::HTTP::Get::new(key)
    res = @ua.request(req)
    return nil if res.code.to_i != 200
    res.body
  end
end

# sample usage
kt = KyotoTycoon::new
kt.open("localhost", 1978)
kt.set("japan", "tokyo", 60)
printf("%s\n", kt.get("japan"))
kt.remove("japan")
kt.close
